lib/linked_list.rb

```diff
@@ -18,122 +18,289 @@ def initialize
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) => where n is the length of the node
+    # Space Complexity: O(1) => because we are not using any extra space
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+
+      until current.nil?
+        puts current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def delete(value)
-      raise NotImplementedError
+    # Time Complexity: O(n) => where n is the length of the node
+    # Space Complexity: O(1) => because we are not using any extra space
+    def delete(value) # 2, [2, 3, 4, 10, 9]
+      return nil if @head.nil?
+      current = @head
+      prev = nil 
+
+      until current.nil?
+        if current.data == value 
+          prev ? prev.next = current.next : @head = current.next
+          return
+        else  
+          prev = current 
+          current = current.next
+        end
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) => where n is the length of the node
+    # Space Complexity: O(1) => because we are not using any extra space
     def reverse
-      raise NotImplementedError
+      current = @head 
+      prev_node = next_node = nil
+
+      while current 
+        next_node = current.next 
+        current.next = prev_node 
+
+        prev_node = current
+        current = next_node 
+      end 
+
+      @head = prev_node
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) => where n is the length of the node
+    # Space Complexity: O(1) => because we are not using any extra space
     def find_middle_value
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      fast_runner = slow_runner = @head
+
+      while !fast_runner.nil? && !fast_runner.next.nil?
+        fast_runner = fast_runner.next.next
+        slow_runner = slow_runner.next
+      end
+
+      return slow_runner.data
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    # Time Complexity: O(n) => where n is the length of the node
+    # Space Complexity: O(1) => because we are not using any extra space
+    # sol1: Use 2 sticks
+    def find_nth_from_end(n) 
+      return nil if @head.nil?
+
+      left_stick = right_stick = @head 
+
+      n.times do 
+        right_stick = right_stick.next
+      end

```

Comment:
You should probably also check to see if `n` is greater than the size of the list, or check to see if `right_stick` is `nil`.